### PR TITLE
#1639 Add the notion of opposite events file for timeline filtering

### PR DESCRIPTION
**Pull request by rosiereflo**

### CODE REVIEW

**Comment by rosiereflo** [comment]
@adriguir FYI

**Comment by rosiereflo** [comment]
With the latest version we need ~5s to filter a timeline with initially 81k lines and reduce it to ~22k lines.
Most of the time is spent in reading the initial timeline.

**Review by valleefred — CHANGES_REQUESTED** [review]

**Review by gautierbureau — APPROVED** [review]

File: dynawo/sources/Common/Dictionaries/validateDictionaries.py [hunk 1]
```diff
@@ -300,6 +308,61 @@ def generate_cpp_files(self, namespace, cpp_dir=None):
             cpp_file.write("};\n")
             cpp_file.write("} //namespace " + namespace + "\n")
 
+    ##
+    # Generate a opposite events file associated to the dictionary
+    # @param self : object pointer
+    # @return
+    def generate_opposite_events_file(self, op_events_dir=None):
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Common/Dictionaries/validateDictionaries.py:316]: solved

File: dynawo/sources/Common/Dictionaries/validateDictionaries.py [hunk 2]
```diff
@@ -300,6 +308,61 @@ def generate_cpp_files(self, namespace, cpp_dir=None):
...
+        if not op_events_dir or len(self.opposite_events) == 0:
+            return
+        name_with_locale = self.name+"_"+self.locale
+        file_name = os.path.join(str(op_events_dir), name_with_locale +'_oppositeEvents.py-tmp')
+        table_to_dump = {}
+        for key1 in self.opposite_events:
+            if not key1 in self.messages:
+                print ("File : DYN"+self.name+"_oppositeEvents.dic, the key " + key1 + " is not found in dictionary file.")
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Common/Dictionaries/validateDictionaries.py]: solved

File: dynawo/sources/Common/Dictionaries/validateDictionaries.py [hunk 3]
```diff
@@ -300,6 +308,61 @@ def generate_cpp_files(self, namespace, cpp_dir=None):
...
+                exit(1)
+            key1_text = self.messages[key1].replace("\'","\\\'")
+            key2 = self.opposite_events[key1]
+            if not key2 in self.messages:
+                print ("File : DYN"+self.name+"_oppositeEvents.dic, the key " + key2 + " is not found in dictionary file.")
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Common/Dictionaries/validateDictionaries.py]: solved

File: dynawo/sources/Common/Dictionaries/validateDictionaries.py [hunk 4]
```diff
@@ -391,11 +454,38 @@ def copy_delete_cpp_header_files(self, cpp_dir=None):
...
 
+    ##
+    # To avoid regeneration of sources, tmp files are created.
+    # If files exist and are equal to tmp files, tmp files are deleted,
+    # else tmp files are copied.
+    #
+    # @param  self: the object pointer
+    # @param  cpp_dir: the path where cpp/header files was created
+    # @return
+    def copy_delete_opposite_event_file(self, cpp_dir=None):
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Common/Dictionaries/validateDictionaries.py]: solved

File: dynawo/sources/Common/test/TestValidateDic.cpp [hunk 5]
```diff
@@ -101,12 +101,20 @@ TEST(Models, TestBuildCheckSum) {
...
+  ASSERT_EQ(result, "Executing command : " + cmd + "File : DYNAIOMyTimeline_oppositeEvents.dic, the key DummyMsg is not found in dictionary file.");
+  ssPython.str(std::string());
+
   cmd = pythonCmd + " validateDictionaries.py --inputDir dic/allInOne/,dic/allInOne/otherDir/share/ "
                     "--outputDir dic/allInOne/ --namespace AIO --existingKeysDir dic/allInOne/otherDir/include/";
   remove("dic/allInOne/AIOMyLog_keys.cpp");
   remove("dic/allInOne/AIOMyLog_keys.h");
   remove("dic/allInOne/AIOMyTimeline_keys.cpp");
   remove("dic/allInOne/AIOMyTimeline_keys.h");
+  remove("dic/allInOne/AIOMyTimeline_oppositeEvents.py");
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Common/test/TestValidateDic.cpp:117]: solved

File: dynawo/sources/Common/CMakeLists.txt [hunk 6]
```diff
@@ -23,15 +23,20 @@ set(DYNAWO_DICTIONARIES
...
+  DYNTimeline_en_GB_oppositeEvents.py
   )
 
-# custom target to generate files associated to dictionnaries
+set(DYNAWO_DICTIONARIES_OPPOSITE_EVENTS
+  Dictionaries/DYNTimeline_oppositeEvents.dic
+  )
+
+# custom target to generate files associated to dictionaries
 add_custom_target(create_keys_files ALL DEPENDS ${DYNAWO_DICTIONARIES})
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Common/CMakeLists.txt:31]: solved

File: util/timeline_filter/timelineFilter.py [hunk 7]
```diff
@@ -140,26 +102,28 @@ def dump(self, filepath, type):
...
         array = line.split('|')
-        if (len(array) != 3):
+        if (len(array) < 3):
             continue
         event = Event()
         time = float(array[0].strip())
         event.time = time
         event.model = array[1].strip()
         event.event = array[2].rstrip().lstrip()
+        if (len(array) ==4):
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply util/timeline_filter/timelineFilter.py]: solved

**Pull request merged**